In [36]:
from shapely.geometry import Point
import contextily as ctx
from shapely.geometry import Point
from typing import List, Union, Set
import os
import requests
import geopandas as gpd
import pandas as pd
import warnings
from sqlalchemy import create_engine
from sqlalchemy.sql import text
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from IPython.display import HTML
warnings.filterwarnings("ignore")

In [2]:

def read_shapefile(shapefile_path: str) -> gpd.GeoDataFrame:
    """
    Reads the shapefile into a GeoDataFrame.

    Args:
    - shapefile_path (str): Path to the shapefile.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame read from the shapefile.
    """
    return gpd.read_file(shapefile_path)

def filter_columns(gdf: gpd.GeoDataFrame, columns: List[str]) -> gpd.GeoDataFrame:
    """
    Filters the GeoDataFrame to include only specified columns.

    Args:
    - gdf (gpd.GeoDataFrame): The original GeoDataFrame.
    - columns (List[str]): A list of column names to retain.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame with only the specified columns.
    """
    return gdf[columns]

def remove_duplicates(gdf: gpd.GeoDataFrame, subset: str) -> gpd.GeoDataFrame:
    """
    Removes duplicate rows based on a specified subset of columns.

    Args:
    - gdf (gpd.GeoDataFrame): The GeoDataFrame to process.
    - subset (str): Column name to check for duplicates.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame with duplicates removed.
    """
    return gdf.drop_duplicates(subset=[subset])

def filter_invalid_zipcodes(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
    """
    Keeps only rows with valid 5-digit zipcodes.

    Args:
    - gdf (gpd.GeoDataFrame): The GeoDataFrame to process.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame with only valid 5-digit zipcodes.
    """
    gdf['zipcode'] = gdf['zipcode'].astype(str)
    return gdf[gdf['zipcode'].str.isdigit() & (gdf['zipcode'].str.len() == 5)]

def process_zipcode_shapefile(shapefile_path: str) -> gpd.GeoDataFrame:
    """
    Cleans and prepares a zipcode shapefile for analysis.

    Args:
    - shapefile_path (str): Path to the zipcode shapefile.

    Returns:
    - gpd.GeoDataFrame: GeoDataFrame with processed zipcode data.
    """
    zipcode_gdf = read_shapefile(shapefile_path)
    essential_columns = ['ZIPCODE', 'geometry']
    zipcode_gdf = filter_columns(zipcode_gdf, essential_columns)
    zipcode_gdf = remove_duplicates(zipcode_gdf, 'ZIPCODE')
    zipcode_gdf.dropna(subset=essential_columns, inplace=True)
    zipcode_gdf.rename(columns={'ZIPCODE': 'zipcode'}, inplace=True)
    zipcode_gdf = filter_invalid_zipcodes(zipcode_gdf)
    common_crs = "EPSG:3857"
    zipcode_gdf.to_crs(common_crs, inplace=True)
    zipcode_gdf.columns = map(str.lower, zipcode_gdf.columns)

    return zipcode_gdf

def lat_validation(latitude):
    if not isinstance(latitude, (int, float)):
        raise TypeError("The latitude should be a float or int type")
    return -90 <= latitude <= 90


def long_validation(longitude: float) -> bool:
    if not isinstance(longitude, (int, float)):
        raise TypeError("The longitude should be a float or int type")
    return -180 <= longitude <= 180



In [3]:
geodf_zip_data = process_zipcode_shapefile("data/nyc_zipcodes/nyc_zipcodes.shp")
nyc_zips = geodf_zip_data['zipcode'].tolist()
nyc_zips = [float(element) for element in nyc_zips]

In [4]:

columns_needed = ['Unique Key', 'Created Date', 'Complaint Type', 'Incident Zip', 'Latitude', 'Longitude', 'Location']
def filter_t311(df: pd.DataFrame, column_needed: List[str], nyc_zip: Union[Set[str], List[str]]) -> gpd.GeoDataFrame:
    # Filter the DataFrame to only include necessary columns and drop rows with NaN values
    filtered = df[column_needed].dropna()

    # Further filter the DataFrame to only include rows where 'Incident Zip' is in nyc_zip
    filtered = filtered[filtered['Incident Zip'].isin(nyc_zip)]

    # Converting 'Created Date' to datetime
    filtered['Created Date'] = pd.to_datetime(filtered['Created Date'])

    # Define your date range
    start_date = pd.to_datetime('2015-01-01')
    end_date = pd.to_datetime('2023-09-30')

    # Filter the DataFrame for dates within the range
    filtered = filtered[(filtered['Created Date'] >= start_date) & (filtered['Created Date'] <= end_date)]

    # Apply latitude and longitude validation
    filtered = filtered[filtered['Latitude'].apply(lat_validation) & filtered['Longitude'].apply(long_validation)]

    # Convert to GeoDataFrame
    filtered = gpd.GeoDataFrame(filtered, geometry=gpd.points_from_xy(filtered['Longitude'], filtered['Latitude']))
    filtered.set_crs("EPSG:4326", inplace=True)
    filtered.to_crs("EPSG:3857", inplace=True)

    return filtered

In [5]:
## cehck
def filter_stc(df: pd.DataFrame, column_needed: List[str], nyc_zip: Set[str]) -> gpd.GeoDataFrame:
    # Filter the DataFrame to only include necessary columns and drop rows with NaN values
    filtered = df[columns_needed].dropna()

    # Further filter the DataFrame to only include rows where 'zipcode' is in nyc_zip
    filtered = filtered[filtered['zipcode'].isin(nyc_zip)]

    # Converting 'created_at' to datetime
    filtered['created_at'] = pd.to_datetime(filtered['created_at'])

    # Define your date range
    start_date = pd.to_datetime('01/01/2015')
    end_date = pd.to_datetime('09/30/2023')  # Corrected date

    # Filter the DataFrame for dates within the range
    filtered = filtered[(filtered['created_at'] >= start_date) & (filtered['created_at'] <= end_date)]

    # Apply latitude and longitude validation
    filtered = filtered[filtered['Latitude'].apply(lat_validation) & filtered['longitude'].apply(long_validation)]

    # Convert to GeoDataFrame
    filtered = gpd.GeoDataFrame(filtered, geometry=gpd.points_from_xy(filtered['longitude'], filtered['Latitude']))
    filtered.set_crs("EPSG:4326", inplace=True)
    filtered.to_crs("EPSG:3857", inplace=True)

    return filtered

In [6]:

def filter_zillow(df: pd.DataFrame, nyc_zip: list) -> pd.DataFrame:
    # Selecting the required columns. Assuming the first column is 'RegionName' and the 9th to last are dates
    useful_cols = df.columns[9:].to_list() + ['RegionName']+['RegionID']
    filtered = df[useful_cols]

    # Drop rows where 'RegionName' is NaN
    filtered = filtered.dropna(subset=['RegionName'])
    filtered = filtered.drop_duplicates()

    # Filter rows where 'RegionName' is in the list of NYC zip codes
    filtered = filtered[filtered['RegionName'].isin(nyc_zip)]
    

    # Melting the DataFrame
    melted_df = filtered.melt(id_vars=['RegionID','RegionName'], value_vars=df.columns[9:], var_name='date', value_name='rent')
    melted_df = melted_df.reset_index()
    return melted_df


In [7]:
a = pd.read_csv("data/zillow_rent_data.csv")
a

RegionID  SizeRank  RegionName RegionType StateName State  \
0        91982         1       77494        zip        TX    TX   
1        91940         3       77449        zip        TX    TX   
2        91733         5       77084        zip        TX    TX   
3        93144         6       79936        zip        TX    TX   
4        62093         7       11385        zip        NY    NY   
...        ...       ...         ...        ...       ...   ...   
6717    418163     30158       89158        zip        NV    NV   
6718     72017     30490       32461        zip        FL    FL   
6719     58956     30490        2876        zip        RI    RI   
6720     91179     30490       76005        zip        TX    TX   
6721     61618     30490       10004        zip        NY    NY   

                   City                                   Metro  \
0                  Katy    Houston-The Woodlands-Sugar Land, TX   
1                  Katy    Houston-The Woodlands-Sugar Land, TX   
2               Houston    Houston-The Woodlands-Sugar Land, TX   
3               El Paso                             El Paso, TX   
4              New York   New York-Newark-Jersey City, NY-NJ-PA   
...                 ...                                     ...   
6717          Las Vegas        Las Vegas-Henderson-Paradise, NV   
6718  Panama City Beach  Crestview-Fort Walton Beach-Destin, FL   
6719   North Smithfield               Providence-Warwick, RI-MA   
6720          Arlington         Dallas-Fort Worth-Arlington, TX   
6721           New York   New York-Newark-Jersey City, NY-NJ-PA   

             CountyName   2015-01-31  ...   2022-12-31   2023-01-31  \
0      Fort Bend County  1606.206406  ...  1994.653463  2027.438438   
1         Harris County  1257.814660  ...  1749.697900  1738.217986   
2         Harris County          NaN  ...  1701.217520  1706.900064   
3        El Paso County          NaN  ...  1419.480272  1458.063897   
4         Queens County          NaN  ...  2935.808220  2895.699421   
...                 ...          ...  ...          ...          ...   
6717       Clark County          NaN  ...  3281.330738  3509.210744   
6718      Walton County          NaN  ...          NaN          NaN   
6719  Providence County          NaN  ...          NaN          NaN   
6720     Tarrant County          NaN  ...  2148.224601  2169.143026   
6721    New York County          NaN  ...  4006.972903  4000.923287   

       2023-02-28   2023-03-31   2023-04-30   2023-05-31   2023-06-30  \
0     2042.237444  2049.325559  2016.531345  2023.438976  2031.558202   
1     1747.305840  1758.407295  1758.891075  1762.980879  1771.751591   
2     1706.067787  1723.722320  1735.484670  1752.132904  1756.990323   
3     1471.726681  1466.734658  1456.175660  1462.478506  1466.267391   
4     2873.209025  2881.906361  2913.546218  2963.964134  3005.735342   
...           ...          ...          ...          ...          ...   
6717  3407.499896  3438.041504  3436.371804  3524.703410  3426.708975   
6718          NaN          NaN          NaN          NaN  2583.675563   
6719          NaN          NaN          NaN          NaN          NaN   
6720  2179.393248  2226.624684  2369.532530  2374.713926  2414.638428   
6721  4002.584212  4085.513015  4224.569373  4240.040733  4286.776061   

       2023-07-31   2023-08-31   2023-09-30  
0     2046.144009  2053.486247  2055.771355  
1     1779.338402  1795.384582  1799.631140  
2     1754.429516  1757.602011  1755.031490  
3     1490.237063  1488.180414  1494.366097  
4     3034.413822  3064.476503  3079.585783  
...           ...          ...          ...  
6717  3412.249969  3310.302151  3448.166667  
6718  2590.977335  2639.938102  2702.500000  
6719          NaN          NaN  2250.000000  
6720  2389.749852  2383.185013  2313.944444  
6721  4270.158740  4353.055657  4355.328283  

[6722 rows x 114 columns]

In [8]:
filter_zillow(a,nyc_zips)

index  RegionID  RegionName        date         rent
0          0     62093       11385  2015-01-31          NaN
1          1     62019       11208  2015-01-31          NaN
2          2     62046       11236  2015-01-31          NaN
3          3     61807       10467  2015-01-31          NaN
4          4     62085       11373  2015-01-31          NaN
...      ...       ...         ...         ...          ...
15325  15325     61773       10282  2023-09-30  7347.458333
15326  15326     62010       11109  2023-09-30  4529.358974
15327  15327     61620       10006  2023-09-30  4060.096154
15328  15328     61723       10162  2023-09-30  5011.666667
15329  15329     61618       10004  2023-09-30  4355.328283

[15330 rows x 5 columns]

In [19]:
def filter_bedbug(df: pd.DataFrame, column_needed: List[str], nyc_zip: Set[str]) -> gpd.GeoDataFrame:
    # Ensure 'Postcode' and 'Filing Date' are in the needed columns
    if 'Postcode' not in column_needed or 'Filing Date' not in column_needed:
        raise ValueError("Required columns 'Postcode' and 'Filing Date' are missing.")

    # Selecting the required columns and drop rows with NaN values
    filtered = df[column_needed].dropna()

    # Further filter the DataFrame to only include rows where 'Postcode' is in nyc_zip
    filtered = filtered[filtered['Postcode'].isin(nyc_zip)]

    # Converting 'Filing Date' to datetime
    filtered['Filing Date'] = pd.to_datetime(filtered['Filing Date'])

    # Define your date range
    start_date = pd.to_datetime('01/01/2015')
    end_date = pd.to_datetime('09/30/2023')

    # Filter the DataFrame for dates within the range
    filtered = filtered[(filtered['Filing Date'] >= start_date) & (filtered['Filing Date'] <= end_date)]
    filtered = filtered.reset_index()
    # Convert to GeoDataFrame (assuming Latitude and Longitude columns are present)
    if 'Latitude' in filtered.columns and 'Longitude' in filtered.columns:
        gdf = gpd.GeoDataFrame(filtered, geometry=gpd.points_from_xy(filtered['Longitude'], filtered['Latitude']))

        # Set a CRS for the GeoDataFrame
        # Replace 'EPSG:3857' with the correct EPSG code to match your PostGIS table
        gdf.set_crs(epsg=3857, inplace=True)
        return gdf
    else:
        raise ValueError("Latitude and Longitude columns are required for GeoDataFrame conversion.")

# Example usage:

In [11]:
# Specify the chunk size
chunk_size = 100000

# Initialize lists to hold processed chunks
t311_chunks = []
stc_chunks = []
zillow_chunks = []
bedbug_chunks = []

# Process and store chunks for '311_Service_Requests'
for chunk in pd.read_csv('data/311_Service_Requests_from_2010_to_Present_20231129.csv', chunksize=chunk_size):
    columns_needed = ['Unique Key', 'Created Date', 'Complaint Type', 'Incident Zip', 'Latitude', 'Longitude', 'Location']
    processed_chunk = filter_t311(chunk, columns_needed, nyc_zip=nyc_zips)
    t311_chunks.append(processed_chunk)
geodf_311_data = pd.concat(t311_chunks)
# Process and store chunks for 'StreetTreesCensus_TREES'
for chunk in pd.read_csv('data/2015StreetTreesCensus_TREES.csv', chunksize=chunk_size):
    columns_needed = ['created_at', 'Latitude', 'longitude', 'tree_id', 'zipcode', 'health', 'spc_common']
    processed_chunk = filter_stc(chunk, columns_needed, nyc_zip=nyc_zips)
    stc_chunks.append(processed_chunk)
geodf_tree_data = pd.concat(stc_chunks)
# Process and store chunks for 'zillow_rent_data'
for chunk in pd.read_csv('data/zillow_rent_data.csv', chunksize=chunk_size):
    processed_chunk = filter_zillow(chunk, nyc_zip=nyc_zips)
    zillow_chunks.append(processed_chunk)
df_zillow_data = pd.concat(zillow_chunks)
# Process and store chunks for 'Bedbug_Reporting'
for chunk in pd.read_csv('data/Bedbug_Reporting_20231203.csv', chunksize=chunk_size):
    columns_needed = ['Building ID', 'Postcode', 'Filing Date', 'Eradicated Unit Count', 'Re-infested  Dwelling Unit Count','Latitude','Longitude']
    processed_chunk = filter_bedbug(chunk, columns_needed, nyc_zip=nyc_zips)
    bedbug_chunks.append(processed_chunk)
df_bedbug_data = pd.concat(bedbug_chunks)

# Load shapefile and save to the database
import geopandas as gpd

# Load your shapefile
geodf_zipcode_data = gpd.read_file('data/nyc_zipcodes/nyc_zipcodes.shp')

# Selecting only the 'ZIPCODE' and 'geometry' columns
geodf_zipcode_data = geodf_zipcode_data[['ZIPCODE', 'geometry']]

# Set the initial CRS to EPSG:2263 if it's not already set
geodf_zipcode_data.set_crs(epsg=2263, inplace=True)

# Transform the CRS from EPSG:2263 to EPSG:3857
geodf_zipcode_data = geodf_zipcode_data.to_crs(epsg=3857)
geodf_zipcode_data = geodf_zipcode_data.drop_duplicates(subset=['ZIPCODE'])


In [ ]:
# Show basic info about each dataframe
geodf_zipcode_data.info()

In [ ]:
# Show first 5 entries about each dataframe
geodf_zipcode_data.head()

In [ ]:
geodf_311_data.info()

In [ ]:
geodf_311_data.head()

In [ ]:
geodf_tree_data.info()

In [ ]:
geodf_tree_data.head()

In [ ]:
df_zillow_data.info()

In [ ]:
df_zillow_data.head()

In [ ]:
df_bedbug_data.info()

In [ ]:
df_bedbug_data.head()

## Part 2: Storing Data

In [ ]:
# this code is not be able to use 
!createdb e4501project
!psql --dbname e4501project -c 'CREATE EXTENSION if NOT EXISTS postgis;'

### Creating Tables


These are just a couple of options to creating your tables; you can use one or the other, a different method, or a combination.

In [213]:
# Define your database connection parameters
db_connection_string = "postgresql://postgres:1234@localhost:5432/e4501project"
engine = create_engine(db_connection_string)

In [13]:
ZIPCODE_SCHEMA = """
CREATE TABLE IF NOT EXISTS nyc_shape (
  "ZIPCODE" float8 PRIMARY KEY,
  "geometry" geometry(POLYGON, 3857)
);
"""

NYC_311_SCHEMA = """
CREATE TABLE IF NOT EXISTS t311 (
    "Unique Key" int8 PRIMARY KEY,
    "Created Date" timestamp(6),
    "Complaint Type" text COLLATE "pg_catalog"."default",
    "Incident Zip" float8,
    "Latitude" float8,
    "Longitude" float8,
    "Location" text COLLATE "pg_catalog"."default",
    "geometry" geometry(POINT, 3857)
);
"""

NYC_TREE_SCHEMA = """
CREATE TABLE IF NOT EXISTS stc (
    "created_at" timestamp(6),
    "Latitude" float8,
    "longitude" float8,
    "tree_id" int8 PRIMARY KEY,
    "zipcode" int8,
    "status" TEXT,
    "health" text COLLATE "pg_catalog"."default",
    "spc_common" text COLLATE "pg_catalog"."default",
    "geometry" geometry(POINT, 3857)
);
"""

ZILLOW_SCHEMA = """
CREATE TABLE IF NOT EXISTS zillow (
  "index" int8 PRIMARY KEY,
  "RegionID" int8,
  "RegionName" int8,
  "date" DATE,
  "rent" float8
)
;
"""

BEDBUG_SCHEMA = '''
CREATE TABLE IF NOT EXISTS Bedbug (
  "index" int   PRIMARY KEY,
  "Building ID" int8,
  "Postcode" float8,
  "Filing Date" timestamp(6),
  "Eradicated Unit Count" float8,
  "Re-infested  Dwelling Unit Count" float8,
  "Latitude" float8,
  "Longitude" float8,
  "geometry" geometry(POINT,3857)
);
'''

In [14]:
# create that required schema.sql file
with open('schema.sql', "w") as f:
    f.write(ZIPCODE_SCHEMA)
    f.write(NYC_311_SCHEMA)
    f.write(NYC_TREE_SCHEMA)
    f.write(ZILLOW_SCHEMA)
    f.write(BEDBUG_SCHEMA)

In [30]:
# Read the SQL schema file
schema_file_path= "schema.sql"
with open(schema_file_path, 'r') as file:
    schema_sql = file.read()

# Execute the SQL schema
with engine.connect() as connection:
    # It's often a good idea to execute commands within a transaction
    with connection.begin() as transaction:
        try:
            connection.execute(schema_sql)
            # Commit the transaction if all commands execute successfully
            transaction.commit()
        except Exception as e:
            # Rollback the transaction in case of an error
            transaction.rollback()
            # Optionally, re-raise the exception or handle it as needed
            raise e

In [16]:
df_zillow_data

index  RegionID  RegionName        date         rent
0          0     62093       11385  2015-01-31          NaN
1          1     62019       11208  2015-01-31          NaN
2          2     62046       11236  2015-01-31          NaN
3          3     61807       10467  2015-01-31          NaN
4          4     62085       11373  2015-01-31          NaN
...      ...       ...         ...         ...          ...
15325  15325     61773       10282  2023-09-30  7347.458333
15326  15326     62010       11109  2023-09-30  4529.358974
15327  15327     61620       10006  2023-09-30  4060.096154
15328  15328     61723       10162  2023-09-30  5011.666667
15329  15329     61618       10004  2023-09-30  4355.328283

[15330 rows x 5 columns]

In [31]:
df_zillow_data.to_sql('zillow', engine, if_exists='append', index=False)
df_bedbug_data.to_postgis('bedbug', engine, if_exists='append', index=False)
geodf_zipcode_data.to_postgis('nyc_shape', engine, if_exists='append', index=False)

In [ ]:
geodf_311_data.to_postgis('t311', engine, if_exists='append', index=False)
geodf_tree_data.to_postgis('stc', engine, if_exists='append', index=False)

## Part 3: Understanding the Data

In [37]:
# Helper function to write the queries to file
import os


def write_query_to_file(query: str, outfile: str) -> None:
    """
    Writes a given query string to a specified file.

    Args:
    query (str): The query string to write to the file.
    outfile (str): The file path where the query will be written.

    Creates the directory for the outfile if it doesn't exist and writes the query to the file.
    """
    # Create the directory for outfile if it doesn't exist
    os.makedirs(os.path.dirname(outfile), exist_ok=True)

    # Open the file and write the query
    with open(outfile, 'w') as file:
        file.write(query)

# Assert test case
test_query = "SELECT * FROM table;"
test_outfile = "/tmp/test_query.sql"

# Execute the function with the test query and outfile
write_query_to_file(test_query, test_outfile)

# Verify the content of the written file
with open(test_outfile, 'r') as file:
    content = file.read()
assert content == test_query, "Test case failed: The content of the file does not match the query"

# Note: This test case writes to a temporary file. Ensure the file path is appropriate for your environment.





### Query 1

In [43]:

QUERY_1_FILENAME = "query/area_more_clam.sql"

QUERY_1 = """
SELECT "Incident Zip", COUNT(*) AS number_of_complaints
FROM t311
WHERE "Created Date" BETWEEN '2022-10-01' AND '2023-09-30'
GROUP BY "Incident Zip"
ORDER BY number_of_complaints DESC;
"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_1))
    for row in result:
        print(row)
write_query_to_file(QUERY_1, QUERY_1_FILENAME)

(11226.0, 48333)
(10467.0, 46368)
(10468.0, 43226)
(10452.0, 43060)
(11385.0, 42863)
(10457.0, 40628)
(10458.0, 40441)
(11207.0, 39236)
(10456.0, 39039)
(11201.0, 38701)
(10453.0, 37411)
(11208.0, 35896)
(10031.0, 33299)
(11221.0, 32466)
(10466.0, 31941)
(10032.0, 31503)
(10025.0, 30499)
(10462.0, 30175)
(11238.0, 29657)
(11230.0, 28901)
(11368.0, 28696)
(11220.0, 28286)
(10019.0, 27730)
(11216.0, 27667)
(11225.0, 27483)
(11235.0, 27258)
(11212.0, 26768)
(10463.0, 26532)
(10460.0, 26469)
(10023.0, 26421)
(11203.0, 26310)
(11214.0, 26141)
(11377.0, 25744)
(11101.0, 25347)
(10027.0, 25327)
(11213.0, 25272)
(10002.0, 25094)
(11211.0, 25044)
(11206.0, 24930)
(10472.0, 24861)
(11234.0, 24845)
(11222.0, 24814)
(11215.0, 24678)
(11209.0, 24671)
(11218.0, 24531)
(11233.0, 24436)
(11223.0, 23549)
(11373.0, 23422)
(10033.0, 23281)
(11237.0, 23012)
(11229.0, 22269)
(11236.0, 21959)
(10034.0, 21868)
(10461.0, 21865)
(10011.0, 21751)
(10003.0, 21241)
(11204.0, 21222)
(10026.0, 20965)
(10469.0, 2077

In conclusion, the area with the zip code 11226 might be more calm to live in, since we found that 11226 has the largest number of 311 complaints in between October 1st, 2022 and September 30th, 2023 (inclusive), which is 48333. 

### Query 2


In [44]:
QUERY_2_FILENAME = "query/most_greenary.sql"

QUERY_2 = """
SELECT zipcode, COUNT(*) AS total_trees
FROM stc
GROUP BY zipcode
ORDER BY total_trees DESC
LIMIT 10;
"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_2))
    for row in result:
        print(row)
write_query_to_file(QUERY_2, QUERY_2_FILENAME)

(10312, 21356)
(10314, 16330)
(10306, 12616)
(10309, 12105)
(11234, 10838)
(11385, 10262)
(11357, 9016)
(11207, 8293)
(11208, 7896)
(11434, 7833)


The 10 areas with the zip codes 10312, 10314, 10306, 10309, 11234, 11385, 11357, 11207, 11208, 11434 have the most trees, which are more greenery than the others.

### Query 3


In [45]:
QUERY_3_FILENAME = "query/most_trees_affordability.sql"

QUERY_3 = """
select zillow."RegionName" , TO_CHAR(AVG(zillow.rent), 'FM999,999,990.00') as avg_price
from zillow inner join (
SELECT zipcode, COUNT(*) AS total_trees
FROM stc
GROUP BY zipcode
ORDER BY total_trees DESC
LIMIT 10
) t on t.zipcode = zillow."RegionName"
where zillow.date >= '2023-08-01'
and zillow.date < '2023-09-01'
group by zillow."RegionName"
order by avg_price desc;

"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_3))
    for row in result:
        print(row)
write_query_to_file(QUERY_3, QUERY_3_FILENAME)

(11207, '3,079.09')
(11385, '3,064.48')
(11208, '2,737.55')
(11434, '2,645.92')
(10314, '2,465.47')
(11357, '2,458.81')
(10306, '2,331.54')
(11234, '2,312.31')
(10309, '1,832.01')
(10312, '1,775.09')


The average rent costs for area with zip code 11207, 11385, 11208, 11434, 10314, 11357, 10306, 10309, 10312 are 3079.09, 3064.48, 2737.55, 2645.92, 2465.47, 2458.81, 2331.54, 2312.31, 1832.01, 1775.09 respectively.

### Query 4


In [46]:
QUERY_4_FILENAME = "query/most_trees_affordability.sql"

QUERY_4 = """
(select zillow."RegionName" , avg(zillow.rent) as avg_price , count(tree_id) as tree_numebr
, count(t311."Unique Key") as compaint_number
from zillow
inner join stc on zillow."RegionName" = stc.zipcode
inner join t311 on t311."Incident Zip" = zillow."RegionName"
where zillow.date >= '2023-01-01'
and zillow.date < '2023-02-01'
and t311."Created Date">= '2023-01-01'
and t311."Created Date"< '2023-02-01'
and stc.created_at< '2023-02-01'
group by zillow."RegionName"
HAVING avg(zillow.rent) is not NULL
order by avg_price desc limit 5)
union
(select zillow."RegionName" , avg(zillow.rent) as avg_price , count(tree_id) as tree_numebr
, count(t311."Unique Key") as compaint_number
from zillow
inner join stc on zillow."RegionName" = stc.zipcode
inner join t311 on t311."Incident Zip" = zillow."RegionName"
where zillow.date >= '2023-01-01'
and zillow.date < '2023-02-01'
and t311."Created Date">= '2023-01-01'
and t311."Created Date"< '2023-02-01'
and stc.created_at< '2023-02-01'
group by zillow."RegionName"
HAVING avg(zillow.rent) is not NULL
order by avg_price ASC limit 5)
"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_3))
    for row in result:
        print(row)
write_query_to_file(QUERY_4, QUERY_4_FILENAME)

(11207, '3,079.09')
(11385, '3,064.48')
(11208, '2,737.55')
(11434, '2,645.92')
(10314, '2,465.47')
(11357, '2,458.81')
(10306, '2,331.54')
(11234, '2,312.31')
(10309, '1,832.01')
(10312, '1,775.09')


For the month of January 2023, the 5 zip codes with the lowest average rent are 10309, 10453, 10462, 11357, 10458 respectively, and 5 zipcodes of the highest average rent are 10013, 10282, 10069, 10011, 10007 respectively, the highest average rents do not correlate directly with the number of trees or complaints.